File: backend/app/models/collection_run.py
from datetime import datetime, timezone
from uuid import uuid4

from sqlalchemy import (
    CheckConstraint,
    Column,
    DateTime,
    Enum,
    Index,
    Integer,
    String,
    Text,
    Uuid,
    text,
)

from app.core.database import Base


COLLECTION_RUN_TYPE_VALUES = (
    "seed_import",
    "runtime_import",
    "collection",
)
COLLECTION_RUN_TRIGGER_TYPE_VALUES = (
    "cli",
    "scheduler",
    "admin",
)
COLLECTION_RUN_STATUS_VALUES = (
    "running",
    "succeeded",
    "partial_failure",
    "failed",
)

COUNT_FIELDS = (
    "requested_count",
    "raw_document_count",
    "extracted_count",
    "accepted_count",
    "partial_count",
    "invalid_count",
    "duplicate_count",
    "rejected_count",
    "inserted_count",
    "updated_count",
    "unchanged_count",
    "skipped_count",
    "failed_count",
)


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


class CollectionRun(Base):
    """Safe execution summary for a Seed, Runtime, or collection run."""

    __tablename__ = "collection_runs"

    run_id = Column(
        Uuid(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )
    source_id = Column(Text, nullable=True)
    run_type = Column(
        Enum(
            *COLLECTION_RUN_TYPE_VALUES,
            name="collection_run_type",
            create_constraint=True,
            validate_strings=True,
        ),
        nullable=False,
    )
    trigger_type = Column(
        Enum(
            *COLLECTION_RUN_TRIGGER_TYPE_VALUES,
            name="collection_run_trigger_type",
            create_constraint=True,
            validate_strings=True,
        ),
        nullable=False,
    )
    started_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=utc_now,
        server_default=text("CURRENT_TIMESTAMP"),
    )
    finished_at = Column(DateTime(timezone=True), nullable=True)
    status = Column(
        Enum(
            *COLLECTION_RUN_STATUS_VALUES,
            name="collection_run_status",
            create_constraint=True,
            validate_strings=True,
        ),
        nullable=False,
        default="running",
        server_default="running",
    )

    requested_count = Column(Integer, nullable=False, default=0, server_default="0")
    raw_document_count = Column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )
    extracted_count = Column(Integer, nullable=False, default=0, server_default="0")
    accepted_count = Column(Integer, nullable=False, default=0, server_default="0")
    partial_count = Column(Integer, nullable=False, default=0, server_default="0")
    invalid_count = Column(Integer, nullable=False, default=0, server_default="0")
    duplicate_count = Column(Integer, nullable=False, default=0, server_default="0")
    rejected_count = Column(Integer, nullable=False, default=0, server_default="0")
    inserted_count = Column(Integer, nullable=False, default=0, server_default="0")
    updated_count = Column(Integer, nullable=False, default=0, server_default="0")
    unchanged_count = Column(Integer, nullable=False, default=0, server_default="0")
    skipped_count = Column(Integer, nullable=False, default=0, server_default="0")
    failed_count = Column(Integer, nullable=False, default=0, server_default="0")

    error_type = Column(String(255), nullable=True)

    __table_args__ = (
        CheckConstraint(
            "source_id IS NULL OR length(trim(source_id)) > 0",
            name="ck_collection_runs_source_id_nonempty",
        ),
        CheckConstraint(
            "finished_at IS NULL OR finished_at >= started_at",
            name="ck_collection_runs_timestamp_order",
        ),
        CheckConstraint(
            "(status = 'running' AND finished_at IS NULL) OR "
            "(status <> 'running' AND finished_at IS NOT NULL)",
            name="ck_collection_runs_terminal_finished_at",
        ),
        CheckConstraint(
            " AND ".join(f"{field} >= 0" for field in COUNT_FIELDS),
            name="ck_collection_runs_counts_nonnegative",
        ),
        Index("ix_collection_runs_source_id", "source_id"),
        Index("ix_collection_runs_started_at", "started_at"),
        Index("ix_collection_runs_status", "status"),
    )

File: frontend/src/types/collectionRun.ts
/**
 * CollectionRun admin API contract.
 *
 * Aligns with docs/api/admin_collection_runs.md and the current OpenAPI schema.
 *
 * List items expose a safe subset; detail includes full count aggregates.
 */

export type CollectionRunType =
  | 'seed_import'
  | 'runtime_import'
  | 'collection';

export type CollectionRunTriggerType = 'cli' | 'scheduler' | 'admin';

export type CollectionRunStatus =
  | 'queued'
  | 'running'
  | 'succeeded'
  | 'partial_failure'
  | 'failed';

export const COLLECTION_RUN_ADMIN_PATH = '/api/v1/admin/collection-runs';

export const COLLECTION_RUN_ADMIN_ENDPOINTS = {
  list: {
    method: 'GET' as const,
    path: COLLECTION_RUN_ADMIN_PATH,
  },
  detail: {
    method: 'GET' as const,
    pathPrefix: `${COLLECTION_RUN_ADMIN_PATH}/`,
  },
  trigger: {
    method: 'POST' as const,
    path: COLLECTION_RUN_ADMIN_PATH,
  },
} as const;

/** List row DTO (`CollectionRunAdminItem`). */
export interface CollectionRunListItemDto {
  run_id: string;
  source_id: string | null;
  run_type: CollectionRunType;
  trigger_type: CollectionRunTriggerType;
  started_at: string;
  finished_at: string | null;
  status: CollectionRunStatus;
  is_stale: boolean;
  is_complete_snapshot: boolean;
  inserted_count: number;
  updated_count: number;
  failed_count: number;
  error_type: string | null;
}

/** Detail DTO (`CollectionRunAdminDetail`). */
export interface CollectionRunDetailDto {
  run_id: string;
  source_id: string | null;
  run_type: CollectionRunType;
  trigger_type: CollectionRunTriggerType;
  started_at: string;
  finished_at: string | null;
  status: CollectionRunStatus;
  is_stale: boolean;
  is_complete_snapshot: boolean;
  requested_count: number;
  raw_document_count: number;
  extracted_count: number;
  accepted_count: number;
  partial_count: number;
  invalid_count: number;
  duplicate_count: number;
  rejected_count: number;
  inserted_count: number;
  updated_count: number;
  unchanged_count: number;
  skipped_count: number;
  failed_count: number;
  error_type: string | null;
}

export interface CollectionRunListQuery {
  page?: number;
  size?: number;
  source_id?: string;
  status?: CollectionRunStatus;
  run_type?: CollectionRunType;
  trigger_type?: CollectionRunTriggerType;
  start_date?: string;
  end_date?: string;
}

export interface CollectionRunListResponse {
  items: CollectionRunListItemDto[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface CollectionRunTriggerRequest {
  source_id?: string;
  requested_count?: number;
}

export interface CollectionRunTriggerResponse {
  run_id: string;
  source_id: string | null;
  run_type: CollectionRunType;
  trigger_type: CollectionRunTriggerType;
  status: CollectionRunStatus;
  started_at: string;
  message: string;
}
